sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt

```diff
@@ -0,0 +1,317 @@
+import kotlin.math.roundToInt
+
+class ReplayIntegration(
+    private val context: Context,
+    private val dateProvider: ICurrentDateProvider
+) : Integration, Closeable, ScreenshotRecorderCallback {
+
+    companion object {
+        const val VIDEO_SEGMENT_DURATION = 5_000L
+        const val VIDEO_BUFFER_DURATION = 30_000L
+    }
+
+    private lateinit var options: SentryOptions
+    private var hub: IHub? = null
+    private var recorder: WindowRecorder? = null
+    private var cache: ReplayCache? = null
+
+    // TODO: probably not everything has to be thread-safe here
+    private val isEnabled = AtomicBoolean(false)
+    private val isRecording = AtomicBoolean(false)
+    private val currentReplayId = AtomicReference<SentryId>()
+    private val segmentTimestamp = AtomicReference<Date>()
+    private val currentSegment = AtomicInteger(0)
+    private val saver =
+        Executors.newSingleThreadScheduledExecutor(ReplayExecutorServiceThreadFactory())
+
+    private val screenBounds by lazy(NONE) {
+        // PixelCopy takes screenshots including system bars, so we have to get the real size here
+        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+        if (VERSION.SDK_INT >= VERSION_CODES.R) {
+            wm.currentWindowMetrics.bounds
+        } else {
+            val screenBounds = Point()
+            @Suppress("DEPRECATION")
+            wm.defaultDisplay.getRealSize(screenBounds)
+            Rect(0, 0, screenBounds.x, screenBounds.y)
+        }
+    }
+
+    private val aspectRatio by lazy(NONE) {
+        screenBounds.bottom.toFloat() / screenBounds.right.toFloat()
+    }
+
+    private val recorderConfig by lazy(NONE) {
+        ScreenshotRecorderConfig(
+            recordingWidth = (720 / aspectRatio).roundToInt(),
+            recordingHeight = 720,
+            scaleFactor = 720f / screenBounds.bottom
+        )
+    }
+
+    override fun register(hub: IHub, options: SentryOptions) {
+        this.options = options
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            options.logger.log(INFO, "Session replay is only supported on API 26 and above")
+            return
+        }
+
+        // TODO: check for replaysSessionSampleRate and replaysOnErrorSampleRate
+
+        this.hub = hub
+        recorder = WindowRecorder(options, recorderConfig, this)
+        isEnabled.set(true)
+    }
+
+    fun isRecording() = isRecording.get()
+
+    fun start() {
+        if (!isEnabled.get()) {
+            options.logger.log(
+                DEBUG,
+                "Session replay is disabled due to conditions not met in Integration.register"
+            )
+            return
+        }
+
+        if (isRecording.getAndSet(true)) {
+            options.logger.log(
+                DEBUG,
+                "Session replay is already being recorded, not starting a new one"
+            )
+            return
+        }
+
+        currentSegment.set(0)
```

Comment:
Should all of this only happen if hub != null?

Reply:
yeah, they are guarded by `if (isEnabled.get())` a few lines above, and that one we only set in `register` after the hub is initialized, so technically should not be the case